src/machine/machine-varlink.c

```diff
@@ -191,13 +194,27 @@ int vl_method_register(sd_varlink *link, sd_json_variant *parameters, sd_varlink
         if (!isempty(machine->root_directory) && path_is_os_tree(machine->root_directory) <= 0)
                 return sd_varlink_error(link, SD_VARLINK_ERROR_INVALID_PARAMETER, NULL);
 
-        /* Ensure an unprivileged user cannot claim any process they don't control as their own machine */
+        /* Ensure an unprivileged user cannot claim any process or directory they don't control as their own machine */
         if (machine->uid != 0) {
                 r = process_is_owned_by_uid(&machine->leader, machine->uid);
                 if (r < 0)
                         return r;
                 if (r == 0)
                         return sd_varlink_error(link, SD_VARLINK_ERROR_PERMISSION_DENIED, NULL);
+
+                if (!isempty(machine->root_directory)) {
+                        _cleanup_free_ char *parent = NULL;
+
+                        r = path_extract_directory(machine->root_directory, &parent);
+                        if (r < 0)
+                                return r;
+
+                        struct stat st;
+                        if (stat(parent, &st) < 0)
+                                return -errno;
+                        if (st.st_uid != machine->uid)
+                                return sd_varlink_error(link, SD_VARLINK_ERROR_PERMISSION_DENIED, NULL);
```

Comment:
I don't follow? Why validate the permission of the parent dir? It's very much reasonable to set up a temporary root tree under /tmp/ or so, no?

Reply:
I guess this could also allow access if the parent is world writable, to cover tmp?